### PR TITLE
Added workflows to trigger table reconciliations

### LINKED ISSUE #927: [FEATURE]: Regression testing harness for post-migration validation for managed tabled (Data Assets)
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

As a customer, much of the risk and work of a large migration is testing and validation. There is currently no *opinionated* pattern from Databricks on how to do regression testing. UC migrations provide an impetus to build one, as this type of migration has high risk.

### Proposed Solution

The feature should:

- Account for any data assets being migrated
- When run, validate post-migration datasets against their pre-migration counterparts. This includes all data assets (managed tables, external tables, views, volumes, ...). See https://github.com/databrickslabs/ucx/issues/906
- Show a view of the outcome of the migration, including success metrics e.g. "% of tables/views/etc successfully migrated", "# of tables migrated". Regression can be classified as dataset mismatch or data asset migration failure. Failures can be detected by capturing failures from the data asset migration tooling in UCX (pending), and tables can be compared against one another to ensure that data and metadata is identical.
- Provide a useful changelog of data and asset changes made that can be analyzed further. Similar to a system table. Example rough schema (includes many optional fields, can be split into denormalized asset-specific tables):

asset_type | hive_database_name | hive_table_name| dbfs_path | uc_metastore_name | uc_schema_name | uc_table_name | view_name | volume_path | migrated_flag | % data migrated

### Additional Context

If any amount of rollback is possible in these kinds of migrations, that should also be considered in this feature's design.

### PR DESCRIPTION
**Pull request by nkvuong**
## Changes

- Added workflows to trigger table reconciliations
- Added widgets to `UCX migration` dashboard displaying reconciliation results

### Linked issues

Resolves #927

### Functionality 

- [x] added relevant user documentation
- [x] added a new workflow `migration_recon_refresh`
- [x] added a new table  `inventory_database.recon_result`

### Tests

- [x] manually tested
- [x] added unit tests
- [x] added integration tests
- [x] verified on staging environment (screenshot attached)

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -458,6 +459,13 @@ There are 3 main table migration workflows, targeting different table types. All
   - Migrate tables using CTAS 
   - Experimentally migrate Delta and Parquet data found in dbfs mount but not registered as Hive Metastore table into UC tables.
 
+### Post Migration Validation Task
+UCX also provides `migrate-recon` workflow to validate the integrity of the migrated tables:
```
Comment (nfx) [inline README.md:463]:
```suggestion
UCX also provides `migrate-data-reconciliation` workflow to validate the integrity of the migrated tables:
```
should we be more explicit about customer-facing interfaces?

File: README.md [hunk 2]
```diff
@@ -458,6 +459,13 @@ There are 3 main table migration workflows, targeting different table types. All
...
+- Comparing the schema of the source and target tables. 
+- Comparing the row counts of the source and target tables.
+- Compare the content of individual row between source and target tables to identify any discrepancies.
+Once completed, the output will be stored in `$inventory_database.recon_result` Delta table.
```
Comment (nfx) [inline README.md:467]: where's the issue for the dashboard to display these results?

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            data_comparison = self._data_comparator.compare_data(source, target)
+            recon_result = ReconResult(
+                schema_comparison.is_matching,
+                (
+                    data_comparison.source_row_count == data_comparison.target_row_count
+                    and data_comparison.num_missing_records_in_target == 0
+                    and data_comparison.num_missing_records_in_source == 0
+                ),
+                schema_comparison.as_dict(),
+                data_comparison.as_dict(),
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:70]: bug: expects a string, but got a dict.

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ReconResult:
+    schema_matches: bool
+    data_matches: bool
+    schema_comparison: str
+    data_comparison: str
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:21]: you probably want to add a view to unfold any nested structures here

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+
+class MigrationRecon(CrawlerBase[ReconResult]):
+    def __init__(
+        self,
+        sbe: SqlBackend,
+        schema,
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:28]: what type it is? :)

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        self._migration_status_refresher.reset()
+        for migration_status in self._migration_status_refresher.snapshot():
+            source = TableIdentifier(
+                "hive_metastore",
+                migration_status.src_schema,
+                migration_status.src_table,
+            )
+            if not self._migration_status_refresher.is_migrated(
+                migration_status.src_schema, migration_status.src_table
+            ):
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:51]: it's a strange logic - first we `refresher.snapshot():` and then we `refresher.is_migrated(..)`. Why don't you get an index and iterate over that?..

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                continue
+            if not migration_status.dst_catalog or not migration_status.dst_schema or not migration_status.dst_table:
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:53]: this condition per se is dupe of previous one

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        self._schema_comparator = schema_comparator
+        self._data_comparator = data_comparator
+
+    def snapshot(self) -> Iterable[ReconResult]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    def _crawl(self) -> Iterable[ReconResult]:
+        self._migration_status_refresher.reset()
+        for migration_status in self._migration_status_refresher.snapshot():
+            source = TableIdentifier(
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:44]: define it closer to the place it's used - 20 lines below

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                continue
+            target = TableIdentifier(
+                migration_status.dst_catalog,
+                migration_status.dst_schema,
+                migration_status.dst_table,
+            )
+            schema_comparison = self._schema_comparator.compare_schema(source, target)
+            data_comparison = self._data_comparator.compare_data(source, target)
+            recon_result = ReconResult(
+                schema_comparison.is_matching,
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:63]: shouldn't we create a thread pool to compute `ReconResult` instances?..

File: src/databricks/labs/ucx/recon/workflows.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+from databricks.labs.ucx.contexts.workflow_task import RuntimeContext
+from databricks.labs.ucx.framework.tasks import Workflow, job_task
+
+
+class MigrationRecon(Workflow):
+    def __init__(self):
+        super().__init__('recon-migration-result')
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/workflows.py:7]:
```suggestion
        super().__init__('data-reconciliation')
```
or something like it

File: src/databricks/labs/ucx/recon/workflows.py [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+    @job_task(job_cluster="main")
+    def recon_migration_result(self, ctx: RuntimeContext):
+        """This workflow validate post-migration datasets against their pre-migration counterparts. This includes all
+        tables, by comparing their schema, row counts and row comparison
+        """
+        ctx.migration_recon.reset()
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/workflows.py:14]: is the reset on an index or something else?

**Review by nfx — CHANGES_REQUESTED** [review]
i think we can benefit a bit from Threads.strict(...) for recon results.

File: src/databricks/labs/ucx/recon/workflows.py (hunk 11 above)
Comment (nkvuong) [reply src/databricks/labs/ucx/recon/workflows.py:14]: need to delete the existing content of recon results table, so that snapshot will re-populate it

File: README.md [hunk 12]
```diff
@@ -458,6 +459,13 @@ There are 3 main table migration workflows, targeting different table types. All
   - Migrate tables using CTAS 
   - Experimentally migrate Delta and Parquet data found in dbfs mount but not registered as Hive Metastore table into UC tables.
 
+### Post Migration Validation Task
+UCX also provides `migrate-data-reconciliation` workflow to validate the integrity of the migrated tables:
+- Comparing the schema of the source and target tables. 
+- Comparing the row counts of the source and target tables.
```
Comment (JCZuurmond) [inline README.md:465]:
```suggestion
- Compare the schema of the source and target tables. 
- Compare the row counts of the source and target tables.
```

Consistent list are easier to read

File: src/databricks/labs/ucx/framework/crawlers.py [hunk 13]
```diff
@@ -52,7 +52,10 @@ def reset(self):
         Delete the content of the inventory table.
         The next call to `snapshot` will re-populate the table.
         """
-        self._exec(f"DELETE FROM {self.full_name}")
+        try:
+            self._exec(f"DELETE FROM {self.full_name}")
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/framework/crawlers.py:56]: Looks out of scope?

File: tests/integration/recon/test_migrate_recon.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+from databricks.sdk.errors import NotFound
+from databricks.sdk.retries import retried
+
+from databricks.labs.ucx.hive_metastore.mapping import Rule
+from databricks.labs.ucx.hive_metastore.tables import What
+
+
+logger = logging.getLogger(__name__)
+
```
Comment (JCZuurmond) [inline tests/integration/recon/test_migrate_recon.py:12]:
```suggestion
```

File: tests/integration/recon/test_migrate_recon.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import logging
```
Comment (JCZuurmond) [inline tests/integration/recon/test_migrate_recon.py:1]:
```suggestion
```

File: tests/integration/recon/test_migrate_recon.py (hunk 14 above)
Comment (JCZuurmond) [reply tests/integration/recon/test_migrate_recon.py:12]: also in other test file

**Review by JCZuurmond — CHANGES_REQUESTED** [review]
Added some nits. 

The most important question to me is how to present the recon result to the users (see comment Serge on a view to unfold nested structures). As all the data is accessible within the workspace, we could show the query for the diff.

File: src/databricks/labs/ucx/framework/crawlers.py (hunk 13 above)
Comment (nkvuong) [reply src/databricks/labs/ucx/framework/crawlers.py:56]: this is needed - if the table does not exist, then reset() would crash

File: src/databricks/labs/ucx/config.py [hunk 16]
```diff
@@ -58,6 +58,9 @@ class WorkspaceConfig:  # pylint: disable=too-many-instance-attributes
     # List of workspace ids ucx is installed on, only applied to account-level installation
     installed_workspace_ids: list[int] | None = None
 
+    # Threshold for row count comparison during data reconciliation, in percentage
+    reconciliation_threshold: int = 5
```
Comment (nfx) [inline src/databricks/labs/ucx/config.py:62]:
```suggestion
    reconciliation_failure_tolerance_percent: int = 5
```

out API guidelines dictate that unit name has to be part of the field name. if it's `_percent`, then probably we need to rename the field a bit as well. just has to sound natural. suggested what _might_ sound natural, pick any other alternative with `_percent` suffix

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 17]
```diff
@@ -27,16 +28,19 @@ class Rule:
     dst_schema: str
     src_table: str
     dst_table: str
+    reconciliation_threshold: int = 0  # threshold for row count comparison
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:31]: missing unit name.

File: src/databricks/labs/ucx/queries/views/reconciliation_results.sql [hunk 18]
```diff
@@ -0,0 +1,26 @@
+with flattened as (
```
Comment (nfx) [inline src/databricks/labs/ucx/queries/views/reconciliation_results.sql:1]:
```suggestion
WITH flattened AS (
```
styleguide: can we use consistently use ALL CAPS for SQL keywords and functions?

File: src/databricks/labs/ucx/recon/base.py [hunk 19]
```diff
@@ -64,13 +65,19 @@ class SchemaComparisonResult:
...
+
 
 @dataclass
 class DataComparisonResult:
     source_row_count: int
     target_row_count: int
-    num_missing_records_in_target: int
-    num_missing_records_in_source: int
+    num_missing_records_in_target: int = 0
+    num_missing_records_in_source: int = 0
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/base.py:77]:
```suggestion
    source_missing_count: int = 0
```
be consistent with units everywhere

File: src/databricks/labs/ucx/recon/base.py [hunk 20]
```diff
@@ -99,7 +106,12 @@ def compare_schema(self, source: TableIdentifier, target: TableIdentifier) -> Sc
 
 class DataComparator(ABC):
     @abstractmethod
-    def compare_data(self, source: TableIdentifier, target: TableIdentifier) -> DataComparisonResult:
+    def compare_data(
+        self,
+        source: TableIdentifier,
+        target: TableIdentifier,
+        row_comparison: bool,
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/base.py:113]:
```suggestion
        compare_rows: bool,
```
might express the intent better, right?..

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+            tasks.append(
+                partial(
+                    self._recon,
+                    source_table,
+                    target_table,
+                    row_comparison,
+                    reconciliation_threshold,
+                )
+            )
+        recon_result, errors = Threads.gather("Reconciling data", tasks)
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:94]:
```suggestion
        recon_result, errors = Threads.gather("reconciling data", tasks)
```

File: src/databricks/labs/ucx/recon/migration_recon.py [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+        if len(errors) > 0:
+            logger.error(f"Detected {len(errors)} while reconciling data")
```
Comment (nfx) [inline src/databricks/labs/ucx/recon/migration_recon.py:96]: can you also show migration errors widget? e.g. we have log errors/warnings widget for assessment, but not for the migration yet.

File: README.md [hunk 23]
```diff
@@ -458,6 +459,13 @@ There are 3 main table migration workflows, targeting different table types. All
   - Migrate tables using CTAS 
   - Experimentally migrate Delta and Parquet data found in dbfs mount but not registered as Hive Metastore table into UC tables.
 
+### Post Migration Data Reconciliation Task
+UCX also provides `migrate-data-reconciliation` workflow to validate the integrity of the migrated tables:
+- Compare the schema of the source and target tables. 
+- Compare the row counts of the source and target tables.
+- Compare the content of individual row between source and target tables to identify any discrepancies.
+Once completed, the output will be stored in `$inventory_database.recon_result` Delta table.
```
Comment (nfx) [inline README.md:467]: can you attach the screenshot with the sample result and short explanation on how to interpret it?..

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by nfx — APPROVED** [review]
Lgtm
